core/base/inc/TPluginManager.h

```diff
@@ -138,38 +143,79 @@ friend class TPluginManager;
    void   SetupCallEnv();
 
    Bool_t CheckForExecPlugin(Int_t nargs);
+   void LoadPluginImpl();
 
 public:
    const char *GetClass() const { return fClass; }
    Int_t       CheckPlugin() const;
    Int_t       LoadPlugin();
 
+   // zero arguments case
+   Longptr_t ExecPluginImpl()
+   {
+      if (!CheckForExecPlugin(0))
+         return 0;
+
+      Longptr_t ret;
+      // locking is handled within this call, but will only be needed
+      // on the first call for initialization
+      fCallEnv->Execute(nullptr, nullptr, 0, &ret);
+
+      return ret;
+   }
+
    template <typename... T> Longptr_t ExecPluginImpl(const T&... params)
    {
-      auto nargs = sizeof...(params);
+      constexpr auto nargs = sizeof...(params);
       if (!CheckForExecPlugin((Int_t)nargs)) return 0;
 
+      Longptr_t ret;
+
+      // check if types match such that function can be called directly
+      const TClass *tupleclass = TClass::GetClass<std::tuple<T...>>();
+      if (tupleclass == fArgTupleClasses[nargs - 1]) {
+         const void *args[nargs] = {&params...};
+         // locking is handled within this call, but will only be needed
+         // on the first call for initialization
+         fCallEnv->Execute(nullptr, args, nargs, &ret);
+
+         return ret;
+      }
+
+      // Fallback to slow path with type conversion for arguments.
+
       // The fCallEnv object is shared, since the PluginHandler is a global
       // resource ... and both SetParams and Execute ends up taking the lock
       // individually anyway ...
 
       R__LOCKGUARD(gInterpreterMutex);
       fCallEnv->SetParams(params...);
 
-      Longptr_t ret;
       fCallEnv->Execute(ret);
 
       return ret;
    }
 
+   // zero arguments case
+   Longptr_t ExecPlugin(int nargs)
+   {
+      // For backward compatibility.
+      if ((gDebug > 1) && (nargs != 0)) {
+         Warning("ExecPlugin", "Announced number of args different from the real number of argument passed %d vs 0",
+                 nargs);
+      }
+
+      return ExecPluginImpl();
+   }
+
    template <typename... T> Longptr_t ExecPlugin(int nargs, const T&... params)
    {
       // For backward compatibility.
       if ((gDebug > 1) && (nargs != (int)sizeof...(params))) {
          Warning("ExecPlugin","Announced number of args different from the real number of argument passed %d vs %lu",
                  nargs, (unsigned long)sizeof...(params) );
       }
-      return ExecPluginImpl(params...);
+      return ExecPluginImpl<T...>(params...);
```

Comment:
Out of curiosity, is the `<T...>` necessary or just for clarity (and the surface it looks 'redundant')?

Reply:
Yes, this preserves the type qualifiers and reference types of the parameter types which are part of the comparison later.

Consider the following case:

```cpp
std::vector<double> v;
pluginmanager->ExecPlugin(v);
```

The template argument deduction will lead to ```ExecPlugin<std::vector<double>>(int, const std::vector<double>&)```
Then we'll get ```ExecPluginImpl<std::vector<double>(const std::vector<double>&)```
and finally ```const TClass *tupleclass = TClass::GetClass<std::tuple<std::vector<double>>>();```

However we might know that for the downstream comparison to match we need instead
```const TClass *tupleclass = TClass::GetClass<std::tuple<const std::vector<double>&>>();```

We can achieve this by calling instead 
```cpp
std::vector<double> v;
pluginmanager->ExecPlugin<const std::vector<double>&>(v);
```

Without the extra <T...> the deduction would give ```ExecPluginImpl<std::vector<double>>(const std::vector<double>&)``` instead of ````ExecPluginImpl<const std::vector<double>&>(const std::vector<double>&)``` as we need.

Reply:
Fair enough.  Thanks.